Link TS generation : splitting into multiple files

guilpier-code opened the pull request:
This PR aims at moving some code about link TS generation, from **main** program to new source files.
It follows PR #2155 

To be done : 

- [x] move headers (*.h) to the right place
- [x] update this branch with its base branch (@flomnes made a change on base branch)

We may also take advantage of this PR to : 
- [ ] Thermal TS generation :
  - [ ] avoid loading the **whole study** when **no** thermal TS generation is required (this is currently the case)
  - [ ] move implementation details about thermal TS generation from **main** program to new source files
- [ ] Separate loading / extracting data from study and generate the TS

guilpier-code committed:
Remove TS links from solver : extract link names from study.

guilpier-code committed:
Remove TS links from solver : extract study general parameters related to link TS generation

guilpier-code committed:
Remove TS links from solver : extract properties from files "properties.ini" for links chosen by user

guilpier-code committed:
Remove TS links from solver : adding the reading of prepro TS

guilpier-code committed:
Remove TS links from solver : actually generating TS for links

guilpier-code committed:
Remove TS links from solver : remove now useless code

guilpier-code committed:
Remove TS links from solver : remove more code + correcting reading errors handling

guilpier-code committed:
Remove TS links from solver : useless code removal

guilpier-code committed:
[skip ci] Remove TS links from solver : code removal

guilpier-code committed:
Remove TS links from solver : review corrections + more simplifications

guilpier-code committed:
Remove TS links from solver : simplifying the data load error handling

guilpier-code committed:
Split TS generation into files : move options handling to separate source files

guilpier-code committed:
Split TS generation into files : start moving code for links TS to new source files

flomnes committed:
Eliminate need for index in loop

guilpier-code committed:
Split TS generation into files : move code related to TS for links from main to other source files

guilpier-code committed:
Split TS generation into files : removing useless header include and namespace declaration

a-zakir commented on src/tools/ts-generator/linksTSgenerator.cpp:
```diff
@@ -0,0 +1,346 @@
+// ==================
+// Free functions
+// ==================
+std::vector<std::string> extractTargetAreas(fs::path sourceLinkDir)
```
const ref is overrated, is it?

guilpier-code replied:
Done

a-zakir commented on src/tools/ts-generator/linksTSgenerator.cpp:
```diff
@@ -0,0 +1,346 @@
+    fs::path pathToIni = sourceLinkDir / "properties.ini";
+    IniFile ini;
+    ini.open(pathToIni); // gp : we should handle reading issues
+    for (auto* s = ini.firstSection; s; s = s->next)
```
auto can handle pointer

guilpier-code replied:
Done

a-zakir commented on src/tools/ts-generator/linksTSgenerator.cpp:
```diff
@@ -0,0 +1,346 @@
+    ini.open(pathToIni); // gp : we should handle reading issues
+    for (auto* s = ini.firstSection; s; s = s->next)
+    {
+        std::string targetAreaName = transformNameIntoID(s->name);
+        to_return.push_back(targetAreaName);
```
```suggestion
        to_return.push_back(transformNameIntoID(s->name));
```

guilpier-code replied:
I did this, because it makes code shorter, but intermediate variable **targetAreaName** had the advantage to be very clear.

guilpier-code replied:
Done

a-zakir commented on src/tools/ts-generator/linksTSgenerator.cpp:
```diff
@@ -0,0 +1,346 @@
+        params.random.reset(seed);
+        return true;
+    }
+    return true; // gp : should we return true here ?
```
indeed, you could log that the key is unknown and therefore skipped

guilpier-code replied:
What you suggest is not what we want : we're reading a **generaldata.ini** file, which (as you know) contains many properties. 
But here, we want to extract some of them. 
If we return **false** as soon as we meet a property we don't want to read, a lot of warnings will be raised.

a-zakir replied:
ok

a-zakir commented on src/tools/ts-generator/linksTSgenerator.cpp:
```diff
@@ -0,0 +1,346 @@
+    std::vector<LinkTSgenerationParams> to_return(linksFromCmdLine.size());
+    // gp : following loop should be improved : we shouldn't need an index
+    unsigned int index = 0;
+    for (const auto& link_pair : linksFromCmdLine)
+    {
+        to_return[index].namesPair = link_pair;
+        index++;
+    }
```
```suggestion
    std::transform(linksFromCmdLine.begin(), linksFromCmdLine.end(), std::back_inserter(to_return), [](const LinkPair& link_pair){
    LinkTSgenerationParams ret;
    ret.namesPair = link_pair;
    return ret;
    } );
```

guilpier-code replied:
I got inspired by your suggestion and eventually used **std::for_each** instead of **std::transform**.
Indeed, in my opinion, **std::for_each** is a bit more adapted to this particular situation, but the idea is the same.

a-zakir commented on src/tools/ts-generator/linksTSgenerator.cpp:
```diff
@@ -0,0 +1,346 @@
+
+std::vector<LinkTSgenerationParams> CreateLinkList(const LinkPairs& linksFromCmdLine)
+{
+    std::vector<LinkTSgenerationParams> to_return(linksFromCmdLine.size());
```
```suggestion
    std::vector<LinkTSgenerationParams> to_return;
```

guilpier-code replied:
Done

a-zakir commented on src/tools/ts-generator/linksTSgenerator.cpp:
```diff
@@ -0,0 +1,346 @@
+std::vector<LinkTSgenerationParams> CreateLinkList(const LinkPairs& linksFromCmdLine)
+{
+    std::vector<LinkTSgenerationParams> to_return(linksFromCmdLine.size());
+    // gp : following loop should be improved : we shouldn't need an index
+    unsigned int index = 0;
```
```suggestion

```

guilpier-code replied:
Done. Use of **index** disappeared.

a-zakir commented on src/tools/ts-generator/linksTSgenerator.cpp:
```diff
@@ -0,0 +1,346 @@
+{
+    for(auto& link : linkList)
+    {
+        if (link.namesPair == link_to_find)
```
why don't you re-use pairs_match?

flomnes replied:
Good point ! `pairs_match` is a bit less strict than `pair<>::operator==`, since it returns `true` if `a.first==b.second && a.second == b.first` (not the case with `pair<>::operator==`)

a-zakir commented on src/tools/ts-generator/linksTSgenerator.cpp:
```diff
@@ -0,0 +1,346 @@
+    }
+}
+
+std::string DateAndTime()
```
there is a function, free from Yuni in [application.cpp](https://github.com/AntaresSimulatorTeam/Antares_Simulator/blob/bce0fa7d3c62bdec6a92aa824d147ddfaec1ffbc/src/solver/application/application.cpp#L429)

guilpier-code replied:
Taken care of.

a-zakir commented on src/tools/ts-generator/linksTSgenerator.cpp:
```diff
@@ -0,0 +1,346 @@
+            std::string sourceAreaName = item.path().filename().generic_string();
+            auto targetAreas = extractTargetAreas(item);
+            for (auto& targetAreaName : targetAreas)
+            {
+                auto linkPair = std::make_pair(sourceAreaName, targetAreaName);
+                to_return.push_back(linkPair);
```
std::transform?

guilpier-code replied:
I think it's clear and short enough as it is.

guilpier-code committed:
Split TS generation into files : simplifying a loop

guilpier-code committed:
Split TS generation into files : move header files to the right place

guilpier-code committed:
Split TS generation into files : taking code review into account

guilpier-code committed:
Split TS generation into files : remove code duplication about time

guilpier-code committed:
Split TS generation into files : renaming

a-zakir commented on src/solver/application/application.cpp:
```diff
@@ -451,7 +437,7 @@ void Application::resetLogFilename() const
 
     logfile /= "solver-";               // append the filename
-    logfile += timeToString() + ".log"; // complete filename with timestamp and extension
+    logfile += FormattedTime("%Y%m%d-%H%M%S") + ".log"; // complete filename with timestamp and extension
```
i'm surprised, how is this works? fs::path+std::string 😮

guilpier-code replied:
`FormattedTime("%Y%m%d-%H%M%S")` is a **std::string**.
So : `FormattedTime("%Y%m%d-%H%M%S") + ".log"` is a **std::string** as well.
Then we use operator **+=** of class **fs::path**, which takes a **std::string** as argument, appends this string to the path.

a-zakir replied:
thanks, i didn't notice the += ope

guilpier-code committed:
Split TS generation into files : simplifying and renaming function writeResultsToDisk which now does exactly what its name says it does

guilpier-code committed:
Split TS generation into files : correction for a regression on CI

guilpier-code committed:
TS : generator : make it work for thermal (#2200)

This PR addresses [this go-pro
ticket](https://gopro-tickets.rte-france.com/browse/ANT-1852)

What was done : 
- **Cause of the crash** : allocate (give a size to) the storage
containers supposed to receive the thermal TS generation results
- Remove the use of **result writer**, too complicated (for example, it
forces us to use a duration collector we don't need).
- Use of **std::filesystem** whenever possible

See helping **gitHub** comments inserted in sources.

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/develop' into fix/split-ts-generation-into-files

guilpier-code committed:
Split TS generation into files : correction after merge

flomnes commented on src/libs/antares/utils/utils.cpp:
```diff
@@ -96,6 +96,20 @@ void BeautifyName(std::string& out, const std::string& oldname)
+    char time_buffer[256];
+    std::strftime(time_buffer, sizeof(time_buffer), format.c_str(), &local_time);
+
+    std::string currentTime = time_buffer;
+
+    return currentTime;
```
```suggestion
    return std::string(time_buffer);
```

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/split-ts-generation-into-files

guilpier-code committed:
TS generator : refactor thermal part (#2202)

Aim of this PR : see the title.

What was done (see helping **gitHub** comments in this PR) : 
- simplifications : 
- [x] Standardizing TS generation with link TS : remove a constructor in
a class containing data for TS generation.
- [x] Remove use of **RuntimeInfos** in **main.cpp** (many code lines
used to be called there, just to get a random generator).
  - [x] Remove check on min stable power, useless.
- [x] Function that actually generates the TS no longer takes the TS to
be generated as argument, but returns it.
- [x] **TS generator** code no longer contains code related to whether
we should skip the TS writing on disk. It writes, that'all. This kind of
code belongs to **solver**, not to **TS generator**.
  - [x] Use **std::filesystem** whenever possible instead of **Yuni**
  
- Unwanted behavior correction : 
- [x] Execute links or thermal TS generation code only depending on user
requirements : indeed, thermal TS generation still needs the whole study
to be loaded, but link TS generation does not (it retrieves from study
only the required data). Being forced to load the study if only link TS
generation is required is a clear performance issue. We avoid that here.
- [x] When data files required to generate TS (modulation files & prepro
files) don't exist, we used to ignore it (this can lead to a crash). We
now raise an error before TS generation time comes.
- Moving code : _to be completed_

---------

Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/split-ts-generation-into-files

flomnes committed:
Formatting

flomnes committed:
Update src/libs/antares/utils/utils.cpp
